File: csharp/perfect-numbers/PerfectNumbers.cs
using System;
using System.Linq;

public enum Classification
{
    Perfect,
    Abundant,
    Deficient
}

public static class PerfectNumbers
{
    public static Classification Classify(int number)
    {
        var aliquot = Enumerable.Range(1, number - 1).Where(o => number / o * o == number).Sum();
        return aliquot == number 
            ? Classification.Perfect 
            : aliquot > number 
                ? Classification.Abundant 
                : Classification.Deficient;
    }
}

File: go/perfect-numbers/perfect_numbers.go
package perfect

import "errors"

// Classification represents a type of perfect number
type Classification int

var (
	// ClassificationDeficient when aliquot sum < number
	ClassificationDeficient = Classification(0)
	// ClassificationPerfect when aliquot sum = number
	ClassificationPerfect = Classification(1)
	// ClassificationAbundant when aliquot sum > number
	ClassificationAbundant = Classification(2)
	// ErrOnlyPositive when given a number less than 1
	ErrOnlyPositive = errors.New("number must be above 0")
)

// Classify returns the classification for a given number
func Classify(n int64) (Classification, error) {
	if n < 1 {
		return 0, ErrOnlyPositive
	}

	aliquot := int64(0)
	for i := int64(1); i < n; i++ {
		if n%i == 0 {
			aliquot += i
		}
	}

	if aliquot < n {
		return ClassificationDeficient, nil
	} else if aliquot == n {
		return ClassificationPerfect, nil
	} else {
		return ClassificationAbundant, nil
	}
}
